src/hmac_key.c

```diff
@@ -0,0 +1,875 @@
+	}
+
+	rv = pthread_mutex_lock(&hmac_key->lock);
+	assert(rv == 0);
+
+	if (hmac_key->refcount != 1) {
+		rc = ZPC_ERROR_OBJINUSE;
+		goto ret;
+	}
+	if (hmac_key->type_set && hmac_key->type == ZPC_HMAC_KEY_TYPE_PVSECRET) {
+		rc = ZPC_ERROR_KEYTYPE;
+		goto ret;
+	}
+	if (hmac_key->hfunc_set != 1) {
+		rc = ZPC_ERROR_HMAC_HASH_FUNCTION_NOTSET;
+		goto ret;
+	}
+
+	memset(&hmac_key->cur, 0, sizeof(hmac_key->cur)); /* no secret ID avail. */
+
+	rc = compute_blocksized_key(key1, key, keylen, hmac_key->hfunc);
+	if (rc != 0) {
+		rc = ZPC_ERROR_CREATE_BLOCKSIZED_KEY;
+		goto ret;
+	}
+
+	rc = hmac_key_clr2prot(hmac_key, key1, hmac_key->keysize / 8);
+	if (rc)
+		goto ret;
+
+	DEBUG("hmac key at %p: key set", hmac_key);
+	hmac_key->key_set = 1;
+	hmac_key->rand_protk = 1;
+
+	rc = 0;
+
+ret:
+	rv = pthread_mutex_unlock(&hmac_key->lock);
+	assert(rv == 0);
+
+	DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+	return rc;
+}
+
+int zpc_hmac_key_export(struct zpc_hmac_key *hmac_key, unsigned char *buf,
+		size_t *buflen)
+{
+	int rc, rv;
+
+	UNUSED(rv);
+
+	if (pkeyfd < 0) {
+		rc = ZPC_ERROR_DEVPKEY;
+		DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+		return rc;
+	}
+	if (hmac_key == NULL) {
+		rc = ZPC_ERROR_ARG1NULL;
+		DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+		return rc;
+	}
+	if (buflen == NULL) {
+		rc = ZPC_ERROR_ARG3NULL;
+		DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+		return rc;
+	}
+
+	rv = pthread_mutex_lock(&hmac_key->lock);
+	assert(rv == 0);
+
+	if (hmac_key->rand_protk) {
+		rc = ZPC_ERROR_PROTKEYONLY;
+		goto ret;
+	}
+
+	rc = hmac_key_check(hmac_key);
+	if (rc)
+		goto ret;
+
+	if (buf == NULL) {
+		*buflen = hmac_key->cur.seclen;
+		rc = 0;
+		goto ret;
+	}
+
+	if (*buflen < hmac_key->cur.seclen) {
+		*buflen = hmac_key->cur.seclen;
+		rc = ZPC_ERROR_SMALLOUTBUF;
+		goto ret;
+	}
+
+	*buflen = hmac_key->cur.seclen;
+	memcpy(buf, hmac_key->cur.sec, *buflen);
+	rc = 0;
+ret:
+	rv = pthread_mutex_unlock(&hmac_key->lock);
+	assert(rv == 0);
+	DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+	return rc;
+}
+
+int zpc_hmac_key_import(struct zpc_hmac_key *hmac_key, const unsigned char *buf,
+		size_t buflen)
+{
+	int rc, rv;
+
+	UNUSED(rv);
+
+	if (pkeyfd < 0) {
+		return ZPC_ERROR_DEVPKEY;
+	}
+	if (hmac_key == NULL) {
+		return ZPC_ERROR_ARG1NULL;
+	}
+	if (buf == NULL) {
+		return ZPC_ERROR_ARG2NULL;
+	}
+	if (buflen != UV_SECRET_ID_LEN) {
+		return ZPC_ERROR_ARG3RANGE;
+	}
+
+	rv = pthread_mutex_lock(&hmac_key->lock);
+	assert(rv == 0);
+
+	if (hmac_key->refcount != 1) {
+		rc = ZPC_ERROR_OBJINUSE;
+		goto ret;
+	}
+	if (hmac_key->keysize_set != 1) {
+		rc = ZPC_ERROR_KEYSIZENOTSET;
+		goto ret;
+	}
+	if (hmac_key->type_set != 1) {
+		rc = ZPC_ERROR_KEYTYPENOTSET;
+		goto ret;
+	}
+	if (hmac_key->type != ZPC_HMAC_KEY_TYPE_PVSECRET) {
+		rc = ZPC_ERROR_KEYTYPE;
+		goto ret;
+	}
+
+	if (hmac_key_blob_is_valid_pvsecret_id(hmac_key, buf) != 0) {
+		rc = ZPC_ERROR_PVSECRET_ID_NOT_FOUND_IN_UV;
+		goto ret;
+	}
+
```

Comment:
Is it checked anywhere that the retrievable secret type is the expected HMAC type? There is ZPC_HMAC_SECRET_HMAC_SHA_256 or ZPC_HMAC_SECRET_HMAC_SHA_512, and the to be imported key jzst match the keysize set in the key object.

Also the hash type set via zpc_hmac_key_set_hash_function() must match to the secret type. Inconsistencies should be rejected (i.e. you can't use hash function SHA-512 with a retrievable secret of type ZPC_HMAC_SECRET_HMAC_SHA_256).

Reply:
Yes, the logic is like:

- setting the hash function for a key obj causes the key size to be set accordingly: 512 or 1024. There is no separate "set_size" function.
- importing a secret ID only checks if the pvsecret is available in the system.
- But when using the key later in the hmac context, the CPACF function code is set depending on the hash function of the key (see local routine __hmac_init in hmac.c). I.e. the hash function is not a property of the context, but only of the key.

Reply:
Right, but what if one set SHA-256 as a hash function, but the secret ID points to a secret of type HMAC_SHA_512. This won't work together. I don't see any checks for that.

Reply:
Per our offline discussions on this, I added commit "Make rc more informative".